src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Internal/MsQuicSession.cs

```diff
@@ -0,0 +1,155 @@
+        private IntPtr _nativeObjPtr;
+        private MsQuicApi _api;
+        private bool _opened;
+
+        internal MsQuicSession(string registration)
+        {
+            _api = new MsQuicApi();
+
+            _api.RegistrationOpen(Encoding.UTF8.GetBytes(registration));
+        }
+
+        public MsQuicConnection ConnectionOpen(IPEndPoint endpoint, SslClientAuthenticationOptions sslClientAuthenticationOptions, IPEndPoint localEndpoint)
+        {
+            if (!_opened)
+            {
+                _nativeObjPtr = _api.SessionOpen(sslClientAuthenticationOptions.ApplicationProtocols[0].Protocol.ToArray());
+                _opened = true;
+                SetPeerBiDirectionalStreamCount(100);
+                SetPeerUnidirectionalStreamCount(100);
+            }
+
+            uint status = _api._connectionOpenDelegate(
+                _nativeObjPtr,
+                MsQuicConnection.NativeCallbackHandler,
+                IntPtr.Zero,
+                out IntPtr connectionPtr);
+
+            MsQuicStatusException.ThrowIfFailed(status);
+
+            return new MsQuicConnection(endpoint, _api, connectionPtr, sslClientAuthenticationOptions);
+        }
+
+        public MsQuicListener ListenerOpen(IPEndPoint listenEndPoint, SslServerAuthenticationOptions sslServerAuthenticationOptions)
+        {
+            if (!_opened)
+            {
+                _nativeObjPtr = _api.SessionOpen(sslServerAuthenticationOptions.ApplicationProtocols[0].Protocol.ToArray());
+                _opened = true;
+                // TODO figure out a story for configuration.
+                SetPeerBiDirectionalStreamCount(100);
+                SetPeerUnidirectionalStreamCount(100);
+            }
+
+            uint status = _api._listenerOpenDelegate(
+                _nativeObjPtr,
+                MsQuicListener.NativeCallbackHandler,
+                IntPtr.Zero,
+                out IntPtr listenerPointer
+                );
+
+            var listener = new MsQuicListener(listenEndPoint, sslServerAuthenticationOptions, _api, listenerPointer);
+
+            return listener;
+        }
+
+        public void ShutDown(
+            QUIC_CONNECTION_SHUTDOWN_FLAG Flags,
+            ushort ErrorCode)
+        {
+            _api._sessionShutdownDelegate(
+                _nativeObjPtr,
+                (uint)Flags,
+                ErrorCode);
+        }
+
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        public void SetPeerBiDirectionalStreamCount(ushort count)
+        {
+            SetUshortParamter(QUIC_PARAM_SESSION.PEER_BIDI_STREAM_COUNT, count);
+        }
+
+        public void SetPeerUnidirectionalStreamCount(ushort count)
+        {
+            SetUshortParamter(QUIC_PARAM_SESSION.PEER_UNIDI_STREAM_COUNT, count);
+        }
+
+        private unsafe void SetUshortParamter(QUIC_PARAM_SESSION param, ushort count)
```

Comment:
You have similar helpers in other classes. I think all common helpers should go into the MsQuicApi to reduce duplication.

Reply:
Good idea.